[analyzer] Prettify checker registration and unittest code

NagyDonat opened the pull request:
This commit tweaks the interface of `CheckerRegistry::addChecker` to make it more practical for plugins and tests:
- The parameter `IsHidden` now defaults to `false` even in the non-templated overload (because setting it to true is unusual, especially in plugins). 
- The parameter `DocsUri` defaults to the dummy placeholder string `"NoDocsUri"` because (as of now) nothing queries its value from the checker registry  (it's only used by the logic that generates the clang-tidy documentation, but that loads it directly from `Checkers.td` without involving the `CheckerRegistry`), so there is no reason to demand specifying this value.

In addition to propagating these changes, this commit clarifies, corrects and extends lots of comments and performs various minor code quality improvements in the code of unit tests and example plugins.

I originally wrote the bulk of this commit when I was planning to add an extra parameter to `addChecker` in order to implement some technical details of the CheckerFamily framework. At the end I decided against adding that extra parameter, so this cleanup was left out of the PR https://github.com/llvm/llvm-project/pull/139256 and I'm merging it now as a separate commit (after minor tweaks).

This commit is mostly NFC: the only functional change is that the analyzer will be compatible with plugins that rely on the default argument values and don't specify `IsHidden` or `DocsUri`. (But existing plugin code will remain valid as well.)

NagyDonat committed:
[analyzer] Prettify checker registration and unittest code

This commit tweaks the interface of `CheckerRegistry::addChecker` to
make it more practical for plugins and tests:
- The most general overload (where `DocsUri` appears in the parameter
  list) is turned into a private method which is only used for loading
  the checkers described in `Checkers.td`. Note that currently _nothing_
  queries the documentation URI from the checker registry (it's only
  used by the logic that generates the clang-tidy documentation, but
  that loads it directly from `Checkers.td` without involving the
  `CheckerRegistry`), so there is no reason to require anyone to provide
  this value. (In fact, it may be a good idea to remove it completely
  from the `CheckerRegistry`.)
- A new public overload is introduced which differs from the most
  general one by omitting the `DocsUri` argument and making the
  `IsHidden` parameter optional (by setting it to `false` by default).
  This will cover the needs of plugins that want maximal customization.
- The templated overload (which takes the checker type as a
  template paarameter + the checker name and description as parameters)
  also loses the `DocsUri` parameter.
- A new method `addMockChecker<T>` is added for use in unit tests.

In addition to propagating these changes (e.g. using `addMockChecker` in
unittests), this commit clarifies, corrects and extends lots of comments
and performs various minor code quality improvements in the code of
unittests and example plugins.

I originally wrote the bulk of this commit when I was planning to add an
extra parameter to `addChecker` in order to implement some technical
details of the CheckerFamily framework. At the end I decided against
adding that extra parameter, so this cleanup was left out of the PR
https://github.com/llvm/llvm-project/pull/139256 and I'm merging it now
as a separate commit (after minor tweaks).

This commit is mostly NFC: the only functional change is that the public
checker registration functions no longer require (or accept) `DocsUri`
as an argument, and the `IsHidden` argument is optional for both public
overloads of `addChecker` (and not just the "basic" templated one).

NagyDonat committed:
Fix a typo

NagyDonat committed:
Fix another typo...

steakhal commented on clang/include/clang/StaticAnalyzer/Frontend/CheckerRegistry.h:
```diff
@@ -112,26 +112,35 @@ class CheckerRegistry {
     addChecker(&CheckerRegistry::initializeManager<CheckerManager, T>,
-               &CheckerRegistry::returnTrue<T>, FullName, Desc, DocsUri,
-               IsHidden);
+               &CheckerRegistry::returnTrue<T>, FullName, Desc,
```
just a random comment: I never understood why does `returnTrue` depend on type `T` if it just returns true xD

steakhal commented on clang/lib/Analysis/plugins/SampleAnalyzer/MainCallChecker.cpp:
```diff
@@ -3,12 +3,16 @@
 class MainCallChecker : public Checker<check::PreStmt<CallExpr>> {
-  mutable std::unique_ptr<BugType> BT;
+
+  BugType BT{this, "call to main", "example analyzer plugin"};
```
Can you make this const?

steakhal reviewed (changes requested):
tiny nits. Indeed it would get simpler :)

NagyDonat committed:
Un-templatify the trivial static method

NagyDonat committed:
Mark the bug type as const

NagyDonat committed:
Clarify doc comments of addChecker overloads

NagyDonat committed:
Satisfy git-clang-format

NagyDonat commented:
@steakhal I implemented the changes that you suggested.

By the way, I'll be on vacation during the next week, so I'll return to this review process only on the 21st of July.

steakhal commented on clang/include/clang/StaticAnalyzer/Frontend/CheckerRegistry.h:
```diff
@@ -108,30 +108,40 @@ class CheckerRegistry {
-  /// Adds a checker to the registry. Use this non-templated overload when your
-  /// checker requires custom initialization.
-  void addChecker(RegisterCheckerFn Fn, ShouldRegisterFunction sfn,
+  /// Adds a checker to the registry.
+  /// This private, most general variant is intended for loading the checker
+  /// definitions from `Checkers.td`.
```
```suggestion
  /// Adds a checker to the registry.
  /// This private, most general overload is intended for loading the checker
  /// definitions from `Checkers.td`.
```

NagyDonat committed:
s/variant/overload/

Co-authored-by: Balazs Benics <benicsbalazs@gmail.com>

NagyDonat committed:
Allow specifying DocsUri

NagyDonat committed:
Remove addMockChecker

NagyDonat committed:
Fix macro parameter interpolation

steakhal reviewed (approved):
Still LGTM.

NagyDonat commented:
@Szelethus I implemented the changes that you suggested in person.

@steakhal FYI I reduced the scope of this change:
- I turned `DocsUri` into an optional argument (with a default value) instead of outright removing it from the `public` overloads.
- I removed `addMockChecker` because @Szelethus suggested that we should not include test-specific utilities in the non-test code.

Szelethus reviewed (approved):
LGTM
